clientv3/watch.go

```diff
@@ -614,13 +603,38 @@ func (w *watchGrpcStream) serveSubstream(ws *watcherStream, resumec chan struct{
 				// shutdown from closeSubstream
 				return
 			}
-			// TODO pause channel if buffer gets too large
-			ws.buf = append(ws.buf, wr)
+
+			skipBuf := wr.Created && !ws.initReq.createdNotify
+			if wr.Created {
+				okToSend := false
+				select {
+				case ws.initReq.retc <- ws.outc:
+					okToSend = true
+				default:
+					// watcher resumed from a disconnect
+				}
+
+				// send creation event only if requested
+				if okToSend && ws.initReq.createdNotify {
+					ws.outc <- *wr
+					skipBuf = true
```

Comment:
@heyitsanthony Yeah that was my concern. I will write more test cases per your suggestion. Thanks!